File: backend/app/services/vnpay_service.py
"""
VNPay Sandbox Integration Service — theo tài liệu chính thức VNPay 2.1.0
https://sandbox.vnpayment.vn/apis/docs/thanh-toan-pay/pay.html

Nguyên tắc tính chữ ký (theo Python sample chính thức của VNPay):
  - Sort params theo alphabetical
  - Build hash string: key=quote_plus(value) nối bằng &
  - Dùng HMAC-SHA512 với HASH_SECRET
  - Hash string và URL query string XÂY DỰNG GIỐNG NHAU (đều quote_plus)
"""
import hashlib
import hmac
import urllib.parse
from datetime import datetime, timedelta, timezone
from typing import Dict, Any
from app.config import settings


VNPAY_RESPONSE_CODES = {
    "00": "Giao dịch thành công",
    "07": "Trừ tiền thành công. Giao dịch bị nghi ngờ gian lận",
    "09": "Chưa đăng ký Internet Banking",
    "10": "Xác thực thông tin không đúng quá 3 lần",
    "11": "Hết thời gian thanh toán",
    "12": "Thẻ/Tài khoản bị khóa",
    "13": "OTP sai",
    "24": "Giao dịch hủy",
    "51": "Tài khoản không đủ số dư",
    "65": "Vượt hạn mức giao dịch trong ngày",
    "75": "Ngân hàng bảo trì",
    "79": "Nhập sai mật khẩu quá số lần quy định",
    "99": "Lỗi khác",
}

# Múi giờ Việt Nam (UTC+7) — VNPay yêu cầu thời gian theo giờ VN
VN_TZ = timezone(timedelta(hours=7))


def _hmac_sha512(key: str, data: str) -> str:
    byte_key = key.encode("utf-8")
    byte_data = data.encode("utf-8")
    return hmac.new(byte_key, byte_data, hashlib.sha512).hexdigest()


def _build_vnpay_query(params: Dict[str, str]) -> str:
    """
    Build query string CHÍNH XÁC theo Python sample chính thức của VNPay:
    - Sort alphabetically
    - Mỗi value được quote_plus (giống urlencode của PHP)
    - Dùng cho cả tính HASH lẫn tạo URL
    """
    hash_data = ""
    i = 0
    for key, val in sorted(params.items()):
        if val and val != "":
            if i == 1:
                hash_data += "&" + key + "=" + urllib.parse.quote_plus(str(val), safe="")
            else:
                hash_data += key + "=" + urllib.parse.quote_plus(str(val), safe="")
            i = 1
    return hash_data


def create_payment_url(
    order_id: int,
    amount: float,
    order_desc: str,
    client_ip: str,
    bank_code: str = "",
    locale: str = "vn",
) -> tuple:
    """
    Build VNPay payment URL với HMAC-SHA512 signature.
    Theo đúng tài liệu VNPay v2.1.0.
    """
    # Thời gian theo giờ Việt Nam (VNPay server ở VN)
    now = datetime.now(VN_TZ)
    vnp_create_date = now.strftime("%Y%m%d%H%M%S")
    vnp_expire_date = (now + timedelta(minutes=15)).strftime("%Y%m%d%H%M%S")

    # VNPay amount = số tiền * 100 (không có dấu thập phân)
    vnp_amount = int(amount * 100)
    vnp_txn_ref = f"{order_id}_{now.strftime('%Y%m%d%H%M%S')}"

    # Chỉ dùng ASCII, không dùng ký tự đặc biệt trong OrderInfo
    safe_desc = order_desc.encode("ascii", errors="ignore").decode("ascii")
    safe_desc = "".join(c for c in safe_desc if c.isalnum() or c in " -_")[:255]

    params: Dict[str, str] = {
        "vnp_Version":   "2.1.0",
        "vnp_Command":   "pay",
        "vnp_TmnCode":   settings.VNPAY_TMN_CODE,
        "vnp_Amount":    str(vnp_amount),
        "vnp_CurrCode":  "VND",
        "vnp_TxnRef":    vnp_txn_ref,
        "vnp_OrderInfo": safe_desc,
        "vnp_OrderType": "other",
        "vnp_Locale":    locale,
        "vnp_ReturnUrl": settings.VNPAY_RETURN_URL,
        "vnp_IpAddr":    client_ip or "127.0.0.1",
        "vnp_CreateDate": vnp_create_date,
        "vnp_ExpireDate": vnp_expire_date,
    }
    if bank_code:
        params["vnp_BankCode"] = bank_code

    ipn = (settings.VNPAY_IPN_URL or "").strip()
    if ipn:
        params["vnp_IpnUrl"] = ipn

    # Tính hash TRÊN CÙNG query string với URL (quote_plus values)
    query_string = _build_vnpay_query(params)
    secure_hash = _hmac_sha512(settings.VNPAY_HASH_SECRET, query_string)

    payment_url = f"{settings.VNPAY_URL}?{query_string}&vnp_SecureHash={secure_hash}"
    return payment_url, vnp_txn_ref


def verify_vnpay_callback(params: Dict[str, str]) -> Dict[str, Any]:
    """
    Verify VNPay callback signature.
    FastAPI tự decode URL params → dùng quote_plus lại để rebuild hash giống VNPay.
    """
    vnp_secure_hash = params.pop("vnp_SecureHash", "")
    params.pop("vnp_SecureHashType", None)

    # Rebuild hash giống cách VNPay tính (quote_plus values, sorted)
    query_string = _build_vnpay_query(params)
    computed_hash = _hmac_sha512(settings.VNPAY_HASH_SECRET, query_string)

    is_valid = hmac.compare_digest(computed_hash.lower(), vnp_secure_hash.lower())
    response_code = params.get("vnp_ResponseCode", "99")
    txn_ref = params.get("vnp_TxnRef", "")
    transaction_id = params.get("vnp_TransactionNo", "")
    amount = int(params.get("vnp_Amount", "0")) / 100  # VNPay gửi x100

    # Lấy order_id từ txn_ref (format: {order_id}_{datetime})
    try:
        order_id = int(txn_ref.split("_")[0]) if "_" in txn_ref else 0
    except (ValueError, IndexError):
        order_id = 0

    return {
        "is_valid": is_valid,
        "is_success": is_valid and response_code == "00",
        "response_code": response_code,
        "response_message": VNPAY_RESPONSE_CODES.get(response_code, "Lỗi không xác định"),
        "order_id": order_id,
        "txn_ref": txn_ref,
        "transaction_id": transaction_id,
        "amount": amount,
        "raw_params": params,
    }

File: backend-dotnet/Services/VnPayService.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using ELearnVN.Backend.Models;
using Microsoft.Extensions.Configuration;

namespace ELearnVN.Backend.Services
{
    public interface IVnPayService
    {
        (string PaymentUrl, string TxnRef) CreatePaymentUrl(int orderId, decimal amount, string orderDesc, string clientIp, string bankCode = "", string locale = "vn");
        VnPayCallbackResult VerifyCallback(Dictionary<string, string> queryParams);
    }

    public class VnPayCallbackResult
    {
        public bool IsValid { get; set; }
        public bool IsSuccess { get; set; }
        public string ResponseCode { get; set; } = null!;
        public string ResponseMessage { get; set; } = null!;
        public int OrderId { get; set; }
        public string TxnRef { get; set; } = null!;
        public string TransactionId { get; set; } = null!;
        public decimal Amount { get; set; }
        public Dictionary<string, string> RawParams { get; set; } = null!;
    }

    public class VnPayService : IVnPayService
    {
        private readonly IConfiguration _config;
        private static readonly Dictionary<string, string> ResponseCodes = new()
        {
            { "00", "Giao dịch thành công" },
            { "07", "Trừ tiền thành công. Giao dịch bị nghi ngờ gian lận" },
            { "09", "Chưa đăng ký Internet Banking" },
            { "10", "Xác thực thông tin không đúng quá 3 lần" },
            { "11", "Hết thời gian thanh toán" },
            { "12", "Thẻ/Tài khoản bị khóa" },
            { "13", "OTP sai" },
            { "24", "Giao dịch hủy" },
            { "51", "Tài khoản không đủ số dư" },
            { "65", "Vượt hạn mức giao dịch trong ngày" },
            { "75", "Ngân hàng bảo trì" },
            { "79", "Nhập sai mật khẩu quá số lần quy định" },
            { "99", "Lỗi khác" }
        };

        public VnPayService(IConfiguration config)
        {
            _config = config;
        }

        public (string PaymentUrl, string TxnRef) CreatePaymentUrl(
            int orderId,
            decimal amount,
            string orderDesc,
            string clientIp,
            string bankCode = "",
            string locale = "vn")
        {
            var tmnCode = _config["VnPay:TmnCode"] ?? "";
            var hashSecret = _config["VnPay:HashSecret"] ?? "";
            var vnpayUrl = _config["VnPay:Url"] ?? "https://sandbox.vnpayment.vn/paymentv2/vpcpay.html";
            var returnUrl = _config["VnPay:ReturnUrl"] ?? "";
            var ipnUrl = _config["VnPay:IpnUrl"] ?? "";

            // Time in Vietnam timezone (UTC+7)
            var utcTime = DateTime.UtcNow;
            var vnTime = TimeZoneInfo.ConvertTimeFromUtc(utcTime, TimeZoneInfo.FindSystemTimeZoneById("SE Asia Standard Time"));
            var createDate = vnTime.ToString("yyyyMMddHHmmss");
            var expireDate = vnTime.AddMinutes(15).ToString("yyyyMMddHHmmss");

            var vnpAmount = (long)(amount * 100); // VNPay requires amount * 100
            var txnRef = $"{orderId}_{vnTime:yyyyMMddHHmmss}";

            // Sanitize order description: keep alphanumeric, space, hyphens, underscores
            var safeDesc = RemoveDiacritics(orderDesc);
            var sb = new StringBuilder();
            foreach (char c in safeDesc)
            {
                if (char.IsLetterOrDigit(c) || c == ' ' || c == '-' || c == '_')
                {
                    sb.Append(c);
                }
            }
            var cleanDesc = sb.ToString();
            if (cleanDesc.Length > 255)
            {
                cleanDesc = cleanDesc.Substring(0, 255);
            }

            var paramsMap = new SortedDictionary<string, string>(StringComparer.Ordinal)
            {
                { "vnp_Version", "2.1.0" },
                { "vnp_Command", "pay" },
                { "vnp_TmnCode", tmnCode },
                { "vnp_Amount", vnpAmount.ToString() },
                { "vnp_CurrCode", "VND" },
                { "vnp_TxnRef", txnRef },
                { "vnp_OrderInfo", cleanDesc },
                { "vnp_OrderType", "other" },
                { "vnp_Locale", locale },
                { "vnp_ReturnUrl", returnUrl },
                { "vnp_IpAddr", string.IsNullOrEmpty(clientIp) ? "127.0.0.1" : clientIp },
                { "vnp_CreateDate", createDate },
                { "vnp_ExpireDate", expireDate }
            };

            if (!string.IsNullOrEmpty(bankCode))
            {
                paramsMap["vnp_BankCode"] = bankCode;
            }

            if (!string.IsNullOrEmpty(ipnUrl))
            {
                paramsMap["vnp_IpnUrl"] = ipnUrl;
            }

            var queryString = BuildQueryString(paramsMap);
            var secureHash = HmacSha512(hashSecret, queryString);
            var paymentUrl = $"{vnpayUrl}?{queryString}&vnp_SecureHash={secureHash}";

            return (paymentUrl, txnRef);
        }

        public VnPayCallbackResult VerifyCallback(Dictionary<string, string> queryParams)
        {
            var hashSecret = _config["VnPay:HashSecret"] ?? "";

            // Clone map and remove hash parameters
            var paramsMap = new SortedDictionary<string, string>(StringComparer.Ordinal);
            string vnpSecureHash = "";

            foreach (var kv in queryParams)
            {
                if (kv.Key == "vnp_SecureHash")
                {
                    vnpSecureHash = kv.Value;
                }
                else if (kv.Key != "vnp_SecureHashType")
                {
                    paramsMap[kv.Key] = kv.Value;
                }
            }

            var queryString = BuildQueryString(paramsMap);
            var computedHash = HmacSha512(hashSecret, queryString);

            bool isValid = string.Equals(computedHash, vnpSecureHash, StringComparison.OrdinalIgnoreCase);

            paramsMap.TryGetValue("vnp_ResponseCode", out var responseCode);
            responseCode ??= "99";

            paramsMap.TryGetValue("vnp_TxnRef", out var txnRef);
            txnRef ??= "";

            paramsMap.TryGetValue("vnp_TransactionNo", out var transactionNo);
            transactionNo ??= "";

            paramsMap.TryGetValue("vnp_Amount", out var amountStr);
            decimal amount = 0;
            if (decimal.TryParse(amountStr, out var amountVal))
            {
                amount = amountVal / 100m;
            }

            int orderId = 0;
            if (!string.IsNullOrEmpty(txnRef) && txnRef.Contains('_'))
            {
                int.TryParse(txnRef.Split('_')[0], out orderId);
            }

            ResponseCodes.TryGetValue(responseCode, out var responseMsg);
            responseMsg ??= "Lỗi không xác định";

            return new VnPayCallbackResult
            {
                IsValid = isValid,
                IsSuccess = isValid && responseCode == "00",
                ResponseCode = responseCode,
                ResponseMessage = responseMsg,
                OrderId = orderId,
                TxnRef = txnRef,
                TransactionId = transactionNo,
                Amount = amount,
                RawParams = queryParams
            };
        }

        private string BuildQueryString(SortedDictionary<string, string> paramsMap)
        {
            var sb = new StringBuilder();
            foreach (var kv in paramsMap)
            {
                if (!string.IsNullOrEmpty(kv.Value))
                {
                    if (sb.Length > 0)
                    {
                        sb.Append("&");
                    }
                    sb.Append(kv.Key);
                    sb.Append("=");
                    sb.Append(VnPayUrlEncode(kv.Value));
                }
            }
            return sb.ToString();
        }

        private string VnPayUrlEncode(string str)
        {
            var sb = new StringBuilder();
            foreach (char c in str)
            {
                if (c == ' ')
                {
                    sb.Append('+');
                }
                else if ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '-' || c == '_' || c == '.' || c == '*')
                {
                    sb.Append(c);
                }
                else
                {
                    sb.Append('%');
                    sb.Append(((int)c).ToString("X2"));
                }
            }
            return sb.ToString();
        }

        private string HmacSha512(string key, string data)
        {
            var keyBytes = Encoding.UTF8.GetBytes(key);
            var dataBytes = Encoding.UTF8.GetBytes(data);
            using var hmac = new HMACSHA512(keyBytes);
            var hashBytes = hmac.ComputeHash(dataBytes);
            return Convert.ToHexString(hashBytes).ToLower();
        }

        private string RemoveDiacritics(string text)
        {
            var normalizedString = text.Normalize(NormalizationForm.FormD);
            var stringBuilder = new StringBuilder();

            foreach (var c in normalizedString)
            {
                var unicodeCategory = CharUnicodeInfo.GetUnicodeCategory(c);
                if (unicodeCategory != UnicodeCategory.NonSpacingMark)
                {
                    stringBuilder.Append(c);
                }
            }

            return stringBuilder.ToString().Normalize(NormalizationForm.FormC);
        }
    }
}
